### ISSUE TITLE
Best place to put email send logs?

### ISSUE THREAD

**Issue by dottjt** [danschultzer/pow#143]
Hey, 

I'm not sure if this is a bit of a random question, but let's say that I want to send out an email log whenever a user gets created. Where would be the best place to put this, given I don't have access to the pow controller without creating my own? 

I was thinking about putting it in the pow routes path, but I don't think that would necessarily work with the changesets. Oh, possibly putting it in the changeset as well, but I'm not entirely sure if that's necessarily correct

Thoughts? Would it make sense maybe to create a hooks module is that we can insert our own code at particular stages of the authentication process?

**Comment by dottjt** [comment]
Actually, I'm thinking the pow mailer might be my best bet. The only issue is that there's no easy way to filter between the emails, unless if I filter by the default subject or text that gets passed through.

**Comment by dottjt** [comment]
It's okay, I've decided to just put it in the pow mailer. Feel free to close this ticket, unless if you think something like function lifecycle hooks would be worth it :)

**Comment by danschultzer** [comment]
Yeah, I definitely want it to be much more explicit than just guessing based on what goes through the email adapter!

I'm planning to redo controller callbacks in #82 which will allow you to hook into the flow with near zero effort. Then it could potentially look like this:

```elixir
defmodule MyAppWeb.PowControllerCallbacks do
   use Pow.Phoenix.ControllerCallbacks

   before_respond Pow.Phoenix.RegistrationController, :create, fn {:ok, user, conn}, ->
     # Send email to admin

     {:ok, user, conn}
   end

   extension PowEmailConfirmation
end
```

The goal is to have something that looks pretty similar to plug, so it's just another conn transformation, and you can hook in everywhere.

However, I've also been thinking in cases like this that it would work well setting up message passing similar to how `:telemetry` works. The context module can be overridden, and you can make the `create/1` method send out a message for the process that can be picked up in your Phoenix app for e-mail delivery. It could look like this:

```elixir
defmodule MyApp.Users do
  use Pow.Ecto.Context,
    repo: MyApp.Repo,
    user: MyApp.Users.User

  def create(params) do
    params
    |> pow_create()
    |> notify_created()
  end

  defp notify_created({:ok, user}) do
    send(self(), {:user_created, user})

    {:ok, user}
  end
  defp notify_created({:error, changeset}), do: {:error, changeset}
end
```

Then receive the message in the Phoenix app for e-mail delivery.

**Comment by dottjt** [comment]
Oh right!

Is that `def create(params) do` already an available feature for pow currently, or just speculating in future how it would work? Because my understanding is that you can only change the changeset for user in terms of the pow user creation pipeline.

Also if you're wondering, since I seem to message you a lot, I basically use pow for my website https://neverfapdeluxe.com, so if you're interested in seeing how it's being implemented there you're more than welcome to take a gander :)

The repo for it is here as well: https://github.com/dottjt/NeverFapDeluxeWebApp

**Comment by danschultzer** [comment]
> Is that def create(params) do already an available feature for pow currently, or just speculating in future how it would work?

You can do that currently, all you have to do is to add `users_context: MyApp.Users` to your config: https://hexdocs.pm/pow/1.0.4/Pow.Ecto.Context.html

I'm not sure what way process messages are best implemented in this case, so I would love to see what your solution is if you decide to go with this.

Your app looks great, and your feedback is very helpful. Real world use cases helps improve Pow a lot 😄

**Comment by dottjt** [comment]
Oh hey, I didn't realise this issue was still open! 

In the end I decided to keep it simple and put it in the `PowMailer` module, although it's really good to know that I have control over the user creation process :)

**Comment by danschultzer** [comment]
Cool! I'm keeping this issue open while working on #82 😄
